### PR TITLE
fix: rework parameter merging rules

### LINKED ISSUE #2173: Concatenate String with Optional Unknown
### Version of Awkward Array

2.0.6

### Description and code to reproduce

I would like to be able to create a `?string` type during concatenation. At it's simplest, this looks like:

```python
import awkward as ak

a = ak.Array([{"a": "foo"}, {"a": "bar"}])
b = ak.Array([{"a": None}])

ak.concatenate([a, b])
```

```pytb
TypeError: while calling

    ak.concatenate(
        arrays = [<Array [{a: 'foo'}, {a: 'bar'}] type='2 * {a: string}'>, <A...
        axis = 0
        mergebool = True
        highlevel = True
        behavior = None
    )

Error details: IndexedOptionArray is not allowed to have parameters["__array__"] = "string"
```

<details>
<summary> Full traceback</summary>

```pytb
File /usr/local/lib/python3.9/site-packages/awkward/operations/ak_concatenate.py:43, in concatenate(arrays, axis, mergebool, highlevel, behavior)
     12 """
     13 Args:
     14     arrays: Array-like data (anything #ak.to_layout recognizes).
   (...)
     31 element for element, and similarly for deeper levels.
     32 """
     33 with ak._errors.OperationErrorContext(
     34     "ak.concatenate",
     35     dict(
   (...)
     41     ),
     42 ):
---> 43     return _impl(arrays, axis, mergebool, highlevel, behavior)

File /usr/local/lib/python3.9/site-packages/awkward/operations/ak_concatenate.py:111, in _impl(arrays, axis, mergebool, highlevel, behavior)
    108         collapsed = ak._do.mergemany(batch)
    109         batch = [ak._do.merge_as_union(collapsed, x)]
--> 111 out = ak._do.mergemany(batch)
    113 if isinstance(out, ak.contents.UnionArray):
    114     out = type(out).simplified(
    115         out._tags,
    116         out._index,
   (...)
    119         mergebool=mergebool,
    120     )

File /usr/local/lib/python3.9/site-packages/awkward/_do.py:304, in mergemany(contents)
    302 def mergemany(contents: list[Content]) -> Content:
    303     assert len(contents) != 0
--> 304     return contents[0]._mergemany(contents[1:])

File /usr/local/lib/python3.9/site-packages/awkward/contents/recordarray.py:635, in RecordArray._mergemany(self, others)
    633 minlength = None
    634 for forfield in for_each_field:
--> 635     merged = forfield[0]._mergemany(forfield[1:])
    637     nextcontents.append(merged)
    639     if minlength is None or merged.length < minlength:

File /usr/local/lib/python3.9/site-packages/awkward/contents/listoffsetarray.py:728, in ListOffsetArray._mergemany(self, others)
    724     return self
    725 listarray = ak.contents.ListArray(
    726     self.starts, self.stops, self._content, parameters=self._parameters
    727 )
--> 728 out = listarray._mergemany(others)
    730 if all(
    731     isinstance(x, ListOffsetArray) and x._offsets.dtype == self._offsets.dtype
    732     for x in others
    733 ):
    734     return out.to_ListOffsetArray64(False)

File /usr/local/lib/python3.9/site-packages/awkward/contents/listarray.py:1084, in ListArray._mergemany(self, others)
   1081 if len(tail) == 0:
   1082     return next
-> 1084 reversed = tail[0]._reverse_merge(next)
   1085 if len(tail) == 1:
   1086     return reversed

File /usr/local/lib/python3.9/site-packages/awkward/contents/indexedoptionarray.py:641, in IndexedOptionArray._reverse_merge(self, other)
    626 self._handle_error(
    627     self._backend[
    628         "awkward_IndexedArray_fill",
   (...)
    637     )
    638 )
    639 parameters = ak._util.merge_parameters(self._parameters, other._parameters)
--> 641 return ak.contents.IndexedOptionArray.simplified(
    642     index, content, parameters=parameters
    643 )

File /usr/local/lib/python3.9/site-packages/awkward/contents/indexedoptionarray.py:124, in IndexedOptionArray.simplified(cls, index, content, parameters)
    117     return IndexedOptionArray(
    118         result,
    119         content.content,
    120         parameters=ak._util.merge_parameters(content._parameters, parameters),
    121     )
    123 else:
--> 124     return cls(index, content, parameters=parameters)

File /usr/local/lib/python3.9/site-packages/awkward/contents/indexedoptionarray.py:59, in IndexedOptionArray.__init__(self, index, content, parameters)
     57 self._index = index
     58 self._content = content
---> 59 self._init(parameters, content.backend)

File /usr/local/lib/python3.9/site-packages/awkward/contents/content.py:69, in Content._init(self, parameters, backend)
     64 else:
     65     if not self.is_list and parameters.get("__array__") in (
     66         "string",
     67         "bytestring",
     68     ):
---> 69         raise ak._errors.wrap_error(
     70             TypeError(
     71                 '{} is not allowed to have parameters["__array__"] = "{}"'.format(
     72                     type(self).__name__, parameters["__array__"]
     73                 )
     74             )
     75         )
     76     if not isinstance(self, ak.contents.NumpyArray) and parameters.get(
     77         "__array__"
     78     ) in ("char", "byte"):
     79         raise ak._errors.wrap_error(
     80             TypeError(
     81                 '{} is not allowed to have parameters["__array__"] = "{}"'.format(
   (...)
     84             )
     85         )

TypeError: while calling

    ak.concatenate(
        arrays = [<Array [{a: 'foo'}, {a: 'bar'}] type='2 * {a: string}'>, <A...
        axis = 0
        mergebool = True
        highlevel = True
        behavior = None
    )

Error details: IndexedOptionArray is not allowed to have parameters["__array__"] = "string"
```

</details>

This operation is not really what I'm trying to do. What I'm really trying to do is:

* https://github.com/scikit-hep/awkward/issues/1668#issuecomment-1238463704
* https://github.com/scikit-hep/awkward/discussions/1647#discussioncomment-3562477

However, this was not an error I ran into last time I tried (using v2 api before 2.0 release). I believe it's intended that this should work, since @jpivarski's suggestion from https://github.com/scikit-hep/awkward/issues/1668#issuecomment-1238463704 now throws the error reported above. E.g.:

```python
a = ak.Array([{"a": "foo"}, {"a": "bar"}])
b = ak.Array([{"b": 1.1}, {"b": 2.2}])

fields = ak.fields(a) + ak.fields(b)

def add_fields(layout, **kwargs):
    if layout.is_record:
        asdict = dict(zip(layout.fields, layout.contents))
        for field in fields:
            if field not in asdict:
                asdict[field] = ak.contents.IndexedOptionArray(
                    ak.index.Index64(np.full(len(layout), -1, np.int64)), ak.contents.EmptyArray()
                )
        return ak.contents.RecordArray(asdict.values(), asdict.keys(), length=len(layout))

a2 = ak.transform(add_fields, a)
b2 = ak.transform(add_fields, b)

ak.concatenate([a2, b2])
```

Now also throws: `Error details: IndexedOptionArray is not allowed to have parameters["__array__"] = "string"`

------------

Is this a bug, or is `?string` no longer a supported type?

If it should be possible, is there a current workaround?

*update: * it looks like `?string` is still a thing: `ak.Array([{"a": "foo"}, {"a": "bar"}, {"a": None}])`, `ak.Array(["foo", None])`

### PR DESCRIPTION
**Pull request by agoose77**
## TL;DR

This PR fixes #2173 by two main changes: make `EmptyArray` a true identity (by ignoring parameters), and consolidate the parameter merging rules.

- Fixes #2173 
- Collapses `_mergeable` and `_mergeable_next` into `_mergeable_next`
- Replace `merge_parameters` with `_parameters_intersect` and `_parameters_union`
- Changes index/option merging rules:
   - option/index — option/index: merge option parameters, merge content parameters
   - option/index — non-option/non-index: keep option parameters, merge content parameters
- Changes union merging rules: 
  - Now always take intersection of merged content parameters
- Changes to union simplification logic:
  - Take union of outermost union parameters and inner union parameters
- Make the base `Content` and `Form` initialiser arguments keyword-only (non-breaking unless you are subclassing)
- Deprecates parameters on empty/unknown types/arrays/forms
- Treats `EmptyArray` as entirely invisible to merging
 

## TL;
This PR is a tentative direction for resolving #2173 by making two big changes:

### Consider `EmptyArray` as truly "empty"
The issue is caused by the fact that `EmptyArray` often has no parameters, which can occur for layouts with only-options, e.g. from Parquet (?) or `ArrayBuilder`. If we try and merge parameters with such layouts, the `merge_equal` rule will lead to the resulting parameter dictionary being empty. This means that we lose `__array__`.

A more permissive rule would fix #2173 is to *only* merge the non-array/record parameters of `EmptyArray`. Such a rule would preserve e.g. strings, but fail to preserve any other parameters whenever a bare `EmptyArray` is encountered. 

I prefer the blanket "EmptyArray-is-empty" approach, but I don't have a good feel for whether that's too restrictive. 

### Index/Option merging rules

In https://github.com/scikit-hep/awkward/discussions/2180 I raise the idea that we need to tighten our parameter rules. This PR makes a particular choice to treat (option, non-option) as merging only the contents. This will make it more obvious that our current logic for `__array__` doesn't look past the top-level node.

### Union merging/simplification

The simple logic followed in this PR is
- `simplified` → union of parameters in simplification
- `merge` → intersection of parameters

Changes to union merging/simplification now reflect this.

### CODE REVIEW

File: src/awkward/contents/bitmaskedarray.py [hunk 1]
```diff
@@ -461,20 +462,18 @@ def _offsets_and_flattened(self, axis, depth):
...
-                ak.contents.IndexedOptionArray,
-                ak.contents.ByteMaskedArray,
-                ak.contents.BitMaskedArray,
-                ak.contents.UnmaskedArray,
-            ),
-        ):
-            return self._content._mergeable(other.content, mergebool)
-
+        # Is the other content is an identity, or a union?
+        if other.is_identity_like or other.is_union:
```
Comment (agoose77) [inline src/awkward/contents/bitmaskedarray.py:466]: This should not be so eager - we care if the `EmptyArray` has explicit parameters

**Comment by agoose77** [comment]
"Parameters don't stick to EmptyArrays."

File: src/awkward/contents/indexedoptionarray.py [hunk 2]
```diff
@@ -1494,18 +1507,24 @@ def _to_backend_array(self, allow_missing, backend):
...
+                elif np.issubdtype(content.dtype.type, np.datetime64) or np.issubdtype(
+                    content.dtype.type, np.timedelta64
+                ):
                     data[mask0] = nplike.asarray(
                         [np.iinfo(np.int64).max], dtype=content.dtype
                     )[0]
+                elif np.issubdtype(content.dtype, np.str_):
+                    data[mask0] = ""
+                elif np.issubdtype(content.dtype, np.bytes_):
+                    data[mask0] = b""
```
Comment (agoose77) [inline src/awkward/contents/indexedoptionarray.py:1527]: Fix masked array of strings

File: src/awkward/contents/unionarray.py [hunk 3]
```diff
@@ -243,16 +249,7 @@ def simplified(
...
-                            contents[k] = (
-                                contents[k]
-                                ._mergemany([inner_cont])
-                                .copy(
-                                    parameters=ak._util.merge_parameters(
-                                        old_parameters, inner_cont._parameters
-                                    )
-                                )
-                            )
+                            contents[k] = contents[k]._mergemany([inner_cont])
```
Comment (agoose77) [inline src/awkward/contents/unionarray.py:257]: I've removed the parameter merging logic here. I don't fully understand _why_ we did things differently before, so I'd appreciate a sense check here @jpivarski

File: src/awkward/_do.py [hunk 4]
```diff
@@ -268,14 +268,20 @@ def num(layout, axis):
...
     theirlength = two.length
-    tags = ak.index.Index8.empty((mylength + theirlength), one.backend.index_nplike)
-    index = ak.index.Index64.empty((mylength + theirlength), one.backend.index_nplike)
+    tags = ak.index.Index8.empty(
+        one.backend.index_nplike.add_shape_item(mylength, theirlength),
+        one.backend.index_nplike,
+    )
+    index = ak.index.Index64.empty(
+        one.backend.index_nplike.add_shape_item(mylength, theirlength),
+        one.backend.index_nplike,
```
Comment (agoose77) [inline src/awkward/_do.py:283]: Unrelated to this PR, but needed fixing

File: src/awkward/contents/unionarray.py (hunk 3 above)
Comment (jpivarski) [reply src/awkward/contents/unionarray.py:257]: I wrote the above code with parameter-merging to make the `simplified` classmethod do what the old `simplify` method used to do. (It was a translation, not a new design.)

`UnionArray.simplify_union` last existed in 1157d228629ff481b7ae26c9653e45655ef11420, and the above line came from

https://github.com/scikit-hep/awkward/blob/1157d228629ff481b7ae26c9653e45655ef11420/src/awkward/contents/unionarray.py#L533

So no, there was no specialized logic here about merging parameters. I was probably thinking "this has to be done because we don't want to lose parameters" but I had forgotten that `_mergemany` merges parameters internally, so I unnecessarily did it another time.

Therefore, it's safe to remove here.

File: src/awkward/contents/listarray.py [hunk 5]
```diff
@@ -943,18 +944,15 @@ def _offsets_and_flattened(self, axis, depth):
...
-            return self._mergeable(other.content, mergebool)
-
+        # Is the other content is an identity, or a union?
+        if other.is_identity_like or other.is_union:
+            return True
+        # Check against option contents
+        elif other.is_option or other.is_indexed:
+            return self._mergeable_next(other.content, mergebool)
+        # Otherwise, do the parameters match? If not, we can't merge.
+        elif not (_type_parameters_equal(self._parameters, other._parameters)):
```
Comment (jpivarski) [inline src/awkward/contents/listarray.py]: PyLint error here.

Actually, this one happens in a few places; just a heads-up.

**Review by jpivarski — CHANGES_REQUESTED** [review]
This is a major piece of work. I like the explicitness of `parameters_intersect` versus `parameters_union`, and the fact that `mergeable` had to be checked before `merge_next` had always been a weak point.

FYI, Parquet can't produce EmptyArray. Only untyped sources.

This PR was motivated by issues that demonstrated output type non-commutativity, and I don't see them in the tests. Let me try to gather them all together.

  1. https://github.com/scikit-hep/awkward/issues/2173#issue-1562867964
  2. https://github.com/scikit-hep/awkward/issues/2173#issuecomment-1409083255
  3. https://github.com/scikit-hep/awkward/discussions/2191#discussioncomment-4852557

Issue (1):

```python
>>> a = ak.Array([{"a": "foo"}, {"a": "bar"}])
>>> b = ak.Array([{"a": None}])
>>> ak.concatenate([a, b])
<Array [{a: 'foo'}, {a: 'bar'}, {a: None}] type='3 * {a: ?string}'>
```

Good!

Issue (2):

```python
>>> from itertools import permutations
>>> string_optional = ak.Array([{"a": "foo"}, {"a": None}])[0:0]
>>> kinds = {"optional": string_optional, "empty": ak.Array([{"a": None}]), "string": ak.Array([{"a": "baz"}])}
>>> for perm in permutations(kinds, 3):
...     print(perm)
...     try:
...         ak.concatenate([kinds[k] for k in perm]).type.show()
...     except TypeError as e:
...         print("errored")
... 
('optional', 'empty', 'string')
2 * {
    a: ?string
}
('optional', 'string', 'empty')
2 * {
    a: ?string
}
('empty', 'optional', 'string')
2 * {
    a: ?string
}
('empty', 'string', 'optional')
2 * {
    a: ?string
}
('string', 'optional', 'empty')
2 * {
    a: ?string
}
('string', 'empty', 'optional')
2 * {
    a: ?string
}
```

Good!

Issue (3):

```python
>>> ak.concatenate((ak.with_parameter([1], "k", "v"), [None])).layout
<IndexedOptionArray len='2'>
    <index><Index dtype='int64' len='2'>[ 0 -1]</Index></index>
    <content><NumpyArray dtype='int64' len='1'>
        <parameter name='k'>'v'</parameter>
        [1]
    </NumpyArray></content>
</IndexedOptionArray>
>>> ak.concatenate(([None], ak.with_parameter([1], "k", "v"))).layout
<IndexedOptionArray len='2'>
    <index><Index dtype='int64' len='2'>[-1  0]</Index></index>
    <content><NumpyArray dtype='int64' len='1'>
        <parameter name='k'>'v'</parameter>
        [1]
    </NumpyArray></content>
</IndexedOptionArray>
```

Good!

So these just need to become formal tests. I'll do that—it'll be quick.

**Comment by jpivarski** [comment]
I had missed the fact that you do have a test, in `tests/test_2179_parameer_merging_rules.py::test_merge_optional_strings` for (1). I'll just add (2) and (3) to that file. (Your other test in that file seems to be something else.)

File: src/awkward/contents/listarray.py (hunk 5 above)
Comment (agoose77) [reply src/awkward/contents/listarray.py]: Do you have this locally? The CI / my local install doesn't see it. It might be that you need to update the pre-commit hooks following Henry's work on ruff.

Comment (jpivarski) [reply src/awkward/contents/listarray.py]: All the PyLint errors seem to be gone. Maybe you fixed them all after my last page-refresh.

**Review by jpivarski — APPROVED** [review]

File: src/awkward/contents/listarray.py (hunk 5 above)
Comment (jpivarski) [reply src/awkward/contents/listarray.py]: I saw the PyLint errors in GitHub, but they're not anymore. (Locally, `pylint` shows a lot more errors through some bad interaction with pyarrow.)
